File: sdk/javascript/src/Network.js
/* eslint-disable no-unused-vars */
import { PublicKey } from './CryptoTypes.js';
import { NetworkTimestamp, NetworkTimestampDatetimeConverter } from './NetworkTimestamp.js';
/* eslint-enable no-unused-vars */
import Ripemd160 from 'ripemd160';

const BASE32_RFC4648_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

/**
 * Represents a network.
 * @template {{bytes: Uint8Array}} TAddress
 * @template {NetworkTimestamp} TNetworkTimestamp
 */
export class Network {
	/**
	 * Creates a new network with the specified name and identifier byte.
	 * @param {string} name Network name.
	 * @param {number} identifier Network identifier byte.
	 * @param {NetworkTimestampDatetimeConverter} datetimeConverter Network timestamp datetime converter associated with this network.
	 * @param {function} addressHasher Gets the primary hasher to use in the public key to address conversion.
	 * @param {function} createAddress Creates an encoded address from an address without checksum and checksum bytes.
	 * @param {AddressConstructable} AddressClass Address class associated with this network.
	 * @param {Constructable} NetworkTimestampClass Network timestamp class associated with this network.
	 */
	constructor(name, identifier, datetimeConverter, addressHasher, createAddress, AddressClass, NetworkTimestampClass) {
		/**
		 * Network name.
		 * @type string
		 */
		this.name = name;

		/**
		 * Network identifier byte.
		 * @type number
		 */
		this.identifier = identifier;

		/**
		 * Network timestamp datetime converter associated with this network.
		 * @type NetworkTimestampDatetimeConverter
		 */
		this.datetimeConverter = datetimeConverter;

		/**
		 * @private
		 */
		this._addressHasher = addressHasher;

		/**
		 * @private
		 */
		this._createAddress = createAddress;

		/**
		 * @private
		 */
		this._AddressClass = AddressClass;

		/**
		 * @private
		 */
		this._NetworkTimestampClass = NetworkTimestampClass;
	}

	/**
	 * Converts a public key to an address.
	 * @param {PublicKey} publicKey Public key to convert.
	 * @returns {TAddress} Address corresponding to the public key input.
	 */
	publicKeyToAddress(publicKey) {
		const partOneHashBuilder = this._addressHasher();
		partOneHashBuilder.update(publicKey.bytes);
		const partOneHash = partOneHashBuilder.digest();

		const partTwoHash = new Ripemd160().update(Buffer.from(partOneHash)).digest();

		const version = new Uint8Array([this.identifier, ...partTwoHash]);

		const partThreeHashBuilder = this._addressHasher();
		partThreeHashBuilder.update(version);
		const checksum = partThreeHashBuilder.digest().subarray(0, 4);

		return this._createAddress(version, checksum);
	}

	/**
	 * Checks if an address string is valid and belongs to this network.
	 * @param {string} addressString Address to check.
	 * @returns {boolean} \c true if address is valid and belongs to this network.
	 */
	isValidAddressString(addressString) {
		if (this._AddressClass.ENCODED_SIZE !== addressString.length)
			return false;

		for (let i = 0; i < addressString.length; ++i) {
			if (-1 === BASE32_RFC4648_ALPHABET.indexOf(addressString[i]))
				return false;
		}

		return this.isValidAddress(new this._AddressClass(addressString));
	}

	/**
	 * Checks if an address is valid and belongs to this network.
	 * @param {TAddress} address Address to check.
	 * @returns {boolean} \c true if address is valid and belongs to this network.
	 */
	isValidAddress(address) {
		if (address.bytes[0] !== this.identifier)
			return false;

		const hashBuilder = this._addressHasher();
		hashBuilder.update(address.bytes.subarray(0, 1 + 20));

		const checkSumFromAddress = address.bytes.subarray(1 + 20);
		const calculatedChecksum = hashBuilder.digest().subarray(0, checkSumFromAddress.length);

		for (let i = 0; i < checkSumFromAddress.length; ++i) {
			if (checkSumFromAddress[i] !== calculatedChecksum[i])
				return false;
		}

		return true;
	}

	/**
	 * Converts a network timestamp to a datetime.
	 * @param {TNetworkTimestamp} referenceNetworkTimestamp Reference network timestamp to convert.
	 * @returns {Date} Datetime representation of the reference network timestamp.
	 */
	toDatetime(referenceNetworkTimestamp) {
		return this.datetimeConverter.toDatetime(Number(referenceNetworkTimestamp.timestamp));
	}

	/**
	 * Converts a datetime to a network timestamp.
	 * @param {Date} referenceDatetime Reference datetime to convert.
	 * @returns {TNetworkTimestamp} Network timestamp representation of the reference datetime.
	 */
	fromDatetime(referenceDatetime) {
		return new this._NetworkTimestampClass(this.datetimeConverter.toDifference(referenceDatetime));
	}

	/**
	 * Returns string representation of this object.
	 * @returns {string} String representation of this object
	 */
	toString() {
		return this.name;
	}
}

/**
 * Provides utility functions for finding a network.
 */
export class NetworkLocator {
	/**
	 * Finds a network with a specified name within a list of networks.
	 * @template {Network<any, any>} TNetwork
	 * @param {Array<TNetwork>} networks List of networks to search.
	 * @param {Array<string>|string} singleOrMultipleNames Names for which to search.
	 * @returns {TNetwork} First network with a name in the supplied list.
	 */
	static findByName(networks, singleOrMultipleNames) {
		const names = Array.isArray(singleOrMultipleNames) ? singleOrMultipleNames : [singleOrMultipleNames];
		const matchingNetwork = networks.find(network => names.some(name => name === network.name));
		if (undefined === matchingNetwork)
			throw RangeError(`no network found with name '${names.join(', ')}'`);

		return matchingNetwork;
	}

	/**
	 * Finds a network with a specified identifier within a list of networks.
	 * @template {Network<any, any>} TNetwork
	 * @param {Array<TNetwork>} networks List of networks to search.
	 * @param {Array<number>|number} singleOrMultipleIdentifiers Identifiers for which to search.
	 * @returns {TNetwork} First network with an identifier in the supplied list.
	 */
	static findByIdentifier(networks, singleOrMultipleIdentifiers) {
		const identifiers = Array.isArray(singleOrMultipleIdentifiers) ? singleOrMultipleIdentifiers : [singleOrMultipleIdentifiers];
		const matchingNetwork = networks.find(network => identifiers.some(identifier => identifier === network.identifier));
		if (undefined === matchingNetwork)
			throw RangeError(`no network found with identifier '${identifiers.join(', ')}'`);

		return matchingNetwork;
	}
}

// region type declarations

/**
 * Constructable class type.
 * @class
 * @typedef {{new(...args: any[]): object}} Constructable
 */

/**
 * Address class type.
 * @class
 * @typedef {{new(...args: any[]): object, ENCODED_SIZE: number}} AddressConstructable
 */

// endregion

File: sdk/python/symbolchain/Network.py
from abc import abstractmethod

from symbolchain.ripemd160 import ripemd160

BASE32_RFC4648_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567'


class Network:
	"""Represents a network."""

	def __init__(self, name, identifier, datetime_converter, address_class, network_timestamp_class):
		# pylint: disable=too-many-arguments

		"""Creates a new network with the specified properties."""
		self.name = name
		self.identifier = identifier
		self.datetime_converter = datetime_converter

		self.address_class = address_class
		self.network_timestamp_class = network_timestamp_class

	def public_key_to_address(self, public_key):
		"""Converts a public key to an address."""
		part_one_hash_builder = self.address_hasher()
		part_one_hash_builder.update(public_key.bytes)
		part_one_hash = part_one_hash_builder.digest()

		part_two_hash = ripemd160(part_one_hash)

		version = bytes([self.identifier]) + part_two_hash

		part_three_hash_builder = self.address_hasher()
		part_three_hash_builder.update(version)
		checksum = part_three_hash_builder.digest()[0:4]

		return self.create_address(version, checksum)

	def is_valid_address_string(self, address_string):
		"""Checks if an address string is valid and belongs to this network."""
		if self.address_class.ENCODED_SIZE != len(address_string):
			return False

		if any(ch not in BASE32_RFC4648_ALPHABET for ch in address_string):
			return False

		return self.is_valid_address(self.address_class(address_string))

	def is_valid_address(self, address):
		"""Checks if an address is valid and belongs to this network."""
		if address.bytes[0] != self.identifier:
			return False

		hash_builder = self.address_hasher()
		hash_builder.update(address.bytes[0:1 + 20])

		checksum_from_address = address.bytes[1 + 20:]
		calculated_checksum = hash_builder.digest()[0:len(checksum_from_address)]
		return checksum_from_address == calculated_checksum

	def to_datetime(self, reference_network_timestamp):
		"""Converts a network timestamp to a datetime."""
		return self.datetime_converter.to_datetime(reference_network_timestamp.timestamp)

	def from_datetime(self, reference_datetime):
		"""Converts a datetime to a network timestamp."""
		return self.network_timestamp_class(self.datetime_converter.to_difference(reference_datetime))

	@abstractmethod
	def address_hasher(self):
		"""Gets the primary hasher to use in the public key to address conversion."""

	@abstractmethod
	def create_address(self, address_without_checksum, checksum):
		"""Creates an encoded address from an address without checksum and checksum bytes."""

	def __eq__(self, other):
		return isinstance(other, Network) and self.name == other.name and self.identifier == other.identifier

	def __str__(self):
		return self.name


class NetworkLocator:
	"""Provides utility functions for finding a network."""

	@staticmethod
	def find_by_name(networks, names):
		"""Finds a network with a specified name within a list of networks."""
		if isinstance(names, str):
			names = [names]

		return next(network for network in networks if network.name in names)

	@staticmethod
	def find_by_identifier(networks, identifiers):
		"""Finds a network with a specified identifier within a list of networks."""
		if isinstance(identifiers, int):
			identifiers = [identifiers]

		return next(network for network in networks if network.identifier in identifiers)
